linkerd/app/src/lib.rs

```diff
@@ -86,7 +86,22 @@ impl Config {
         debug!("building app");
         let (metrics, report) = Metrics::new(admin.metrics_retain_idle);
 
-        let dns = info_span!("dns").in_scope(|| dns.build())?;
+        // Because constructing a `trust-dns` resolver is an async fn that
+        // spawns the DNS background task on the runtime that executes it,
+        // wehave to construct the admin runtime early and use it to build the
+        // DNS resolver. When we spawn the admin thread, we will move the
+        // runtime constructed here to that thread and have it execute the admin
+        // workloads.
+        let mut admin_rt = tokio_02::runtime::Builder::new()
+            .basic_scheduler()
+            .enable_all()
+            .build()
+            .expect("admin runtime");
+
+        let dns = dns
+            .build(admin_rt.handle().clone())
+            .instrument(info_span!("dns"))
+            .await;
```

Comment:
WDYT about putting this in `admin::Config::build` and then just giving a handle to the resolver off of that?

Definitely not a blocker.

Reply:
Hmm, that seemed nice at first until I actually tried it, and it turned out to be kinda awkward: we currently construct the identity client and then pass it in to `admin::Config::build`, and constructing the identity client depends on the DNS resolver:
https://github.com/linkerd/linkerd2-proxy/pull/520/files#diff-104ba36cb3e84114b06ebde95af46477R107

so, I _think_ that if we were going to have the `Admin` struct own the runtime, we'd need to make the resolver lazy internally?

Reply:
Nevermind. Let's come back to this later.